exasol/driver/websocket.py

```diff
@@ -0,0 +1,556 @@
+"""
+This module provides a `PEP-249 <https://peps.python.org/pep-0249/#interfaceerror>`_ compliant DBAPI interface,
+for a `websocket based <https://github.com/exasol/websocket-api>`_ `EXASOL <https://www.exasol.com/de/>`_ database driver.
+"""
+from datetime import (
+    date,
+    datetime,
+    time,
+)
+from enum import (
+    IntFlag,
+    auto,
+)
+from functools import wraps
+from time import localtime
+from typing import Protocol
+
+import pyexasol
+
+apilevel = "2.0"
+threadsafety = 1
+paramstyle = "qmark"
+
+
+class Warning(Exception):
+    """Exception raised for important warnings like data truncations while inserting, etc."""
+
+
+class Error(Exception):
+    """
+    Exception that is the base class of all other error exceptions.
+    You can use this to catch all errors with one single except statement.
+    Warnings are not considered errors and thus should not use this class as base.
+    """
+
+
+class InterfaceError(Error):
+    """Exception raised for errors that are related to the database interface rather than the database itself."""
+
+
+class DatabaseError(Error):
+    """Exception raised for errors that are related to the database."""
+
+
+class DataError(DatabaseError):
+    """
+    Exception raised for errors that are due to problems with the processed data like division by zero,
+    numeric value out of range, etc."""
+
+
+class OperationalError(DatabaseError):
+    """
+    Exception raised for errors that are related to the database’s operation
+    and not necessarily under the control of the programmer, e.g. an unexpected disconnect occurs,
+    the data source name is not found, a transaction could not be processed,
+    a memory allocation error occurred during processing, etc.
+    """
+
+
+class IntegrityError(DatabaseError):
+    """
+    Exception raised when the relational integrity of the database is affected,
+    e.g. a foreign key check fails.
+    """
+
+
+class InternalError(DatabaseError):
+    """
+    Exception raised when the database encounters an internal error,
+    e.g. the cursor is not valid anymore, the transaction is out of sync, etc.
+    """
+
+
+class ProgrammingError(DatabaseError):
+    """
+    Exception raised for programming errors, e.g. table not found or already exists,
+    syntax error in the SQL statement, wrong number of parameters specified, etc.
+    """
+
+
+class NotSupportedError(DatabaseError):
+    """
+    Exception raised in case a method or database API was used which is not supported by the database,
+    e.g. requesting a .rollback() on a connection that does not support transaction
+    or has transactions turned off.
+    """
+
+
+class Connection(Protocol):
+    """
+    Defines a protocol which is compliant with https://peps.python.org/pep-0249/#connection-objects connection objects.
```

Comment:
Based on the class name "Connection" my assumption would be that this class defines a _connection_  rather than a "protocol" or do I oversee something? 
```suggestion
    Defines a connection based on the specified protocol which complies to https://peps.python.org/pep-0249/#connection-objects.
```